[장바구니 협업 Step 2] 세인(이세민) 미션 제출합니다.

semnil5202 opened the pull request:
안녕하세요 카일! 그간 잘 지내셨나요?? 😀

이번 미션 정말 쉽지 않았던 것 같습니다.. 구현 하는데만 시간이 벅차 MSW, 스토리북, 테스트 등에 전혀 신경을 못 썼습니다.

구현한 기능에 대해서 설명을 드리겠습니다.

저희 팀은 **쿠폰** 재화 시스템을 선택하였고, 로그인 및 회원가입 기능이 존재합니다.

쿠폰을 영구적으로 사용할 수 있게하면 로그인 및 회원가입 기능이 불필요하지만, 백엔드의 요구사항 중 하나가 **현실 세계를 반영하여** 재화 시스템을 설계해야한다는 것이었기에 쿠폰을 한정적으로 사용할 수 밖에 없었습니다.

카일이 사용하실 때 쿠폰이 만료 되어 사용을 하지 못할 수도 있으므로 로그인 및 회원가입 기능을 추가하게 되었습니다.

따라서 회원가입 후 로그인하여 홈페이지를 사용하시면 감사드리겠습니다. 😀

### 컴포넌트 설계
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/0ffe634b-c346-4b5b-b573-4c0c52b17f5f)
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/437d84f3-8486-42ab-94c3-9858c815a2e9)
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/8b2f20a7-5fde-4746-9a06-f2e7ee8b3e54)
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/94f9abd6-8152-494c-93b6-0d810f34d05a)
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/70386dc0-22e4-4ba0-b019-e782ceaf1e56)
![image](https://github.com/woowacourse/react-shopping-cart-prod/assets/89172499/59aab2e4-501f-4979-9da1-2c2f9b3290b8)

### 궁금한 점
#### Q1 : 상태가 점점 많아지고 복잡해지는 것 같은데, 리팩토링 방향을 어떻게 잡을지 고민입니다.
구현을 하는 중간 중간 특정 상태가 전역적으로 관리될 필요가 있다면, 그때 그때 useState에서 Recoil로 옮겼습니다. 그러다보니 상태를 두서 없이 관리하고 있다는 느낌이 강합니다. 카일이 보시기에 지금의 상태 관리가 괜찮은 수준인지, 어떻게 개선해야할 지 말씀해주시면 감사하겠습니다!

semnil5202 committed:
chore: 디렉토리, 스토리북 및 스타일 로직 분리

semnil5202 committed:
refactor(Toast): Toast가 화면에 유지되는 시간 변경(1.6s -> 2s)

semnil5202 committed:
fix(ProductList): 스켈레톤 UI key prop 오류 해결

스켈레톤 UI는 일회성 및 변경 가능성이 거의 없으므로 index로 key 값을 지정한다.

semnil5202 committed:
refactor: Logo 동적 변경 기능 추가

기존 Header 컴포넌트에서 Logo부분을 분리한다.
각각의 서버마다 장바구니 정보가 다르므로 서버 이름이 장바구니 페이지 이름이 될 수 있도록한다.

semnil5202 committed:
refactor: 서버네임 타입 값 변경 (한글 -> 영어)

semnil5202 committed:
refactor: 전반적인 페이지 사이즈 및 반응형 CSS 수정

semnil5202 committed:
refactor(ProductList): 상품 정보 에러를 먼저 보여주는 기능 추가

상품 정보와 장바구니 정보를 모두 불러오지 못했을 경우 상품 정보 에러만 보여주도록 한다.

semnil5202 committed:
refactor: 상품 수량 수정 버튼 스타일 수정

semnil5202 committed:
docs: 기능 구현 목록 작성

semnil5202 committed:
refactor: serverName Atom 분리

semnil5202 committed:
refactor: toast Atom 분리

semnil5202 committed:
refactor: cart Atom 명 변경(state -> cart)

semnil5202 committed:
feat: 모달 컴포넌트 추가

semnil5202 committed:
feat: 중복되지 않은 아이디 및 패스워드를 통해 회원가입을 하는 기능 추가

semnil5202 committed:
feat: 회원가입한 아이디 및 패스워드로 로그인 하는 기능 추가

semnil5202 committed:
refactor: 로그인 및 회원가입 api prefix 변경 (submit -> post)

semnil5202 committed:
refactor: Authorization 정책 변경으로 인한 API 로직 수정

semnil5202 committed:
refactor: 불필요한 console.log 제거

semnil5202 committed:
feat: 첫 로그인과 동시에 신규 가입 쿠폰을 받는 기능 및 쿠폰 구분 기능 추가

semnil5202 committed:
refactor: Primary Color CSS 변수 지정

semnil5202 committed:
chore: OS 차이로 인한 NODE_ENV 에러 대응 위해 cross-env 추가

semnil5202 committed:
feat: 사용한 쿠폰은 만료처리 하는 기능 추가

추가로 쿠폰 스타일 값을 수정하였다.

semnil5202 committed:
refactor: 기본 서버 설정 변경 (RAON -> JOURNY)

semnil5202 committed:
fix: 비제어 컴포넌트 오류 수정 (readOnly 속성 지정)

semnil5202 committed:
refactor: cursor 옵션 제거

not allowed는 과한 조치라고 판단하여 제거한다.

semnil5202 committed:
feat: 결제예상금액에서 쿠폰 선택 및 주문 기능 추가

semnil5202 committed:
feat: 상품 페이지에서 Coupon 데이터 받아오는 로직 및 핸들링추가

semnil5202 committed:
feat: 로그아웃 시 로그인 토큰, 쿠폰함, 장바구니 상태 및 로컬스토리지 비우는 기능 추가

semnil5202 committed:
chore: 맥북 OS 환경에서 cross-env 제거

semnil5202 committed:
refactor: 불필요한 console.log 제거

semnil5202 committed:
feat: 주문한 상품은 장바구니에서 제거하는 기능

semnil5202 committed:
feat: 쿠폰 할인률에 따라 할인 받을 수 있는 금액 보여주는 기능 추가

semnil5202 committed:
feat: 결제한 주문건에 대하여 주문목록을 볼 수 있는 기능 추가

semnil5202 committed:
feat: 사용자가 주문한 건에 대해서 상세 물품 정보를 확인할 수 있는 기능 추가

semnil5202 committed:
feat: 상세보기 버튼을 눌러 주문상세내역 페이지로 이동하는 기능 추가

semnil5202 committed:
feat: 쿠폰 할인 금액은 primary color로 표시

semnil5202 committed:
docs: 누락된 기능 구현 목록 수정

semnil5202 committed:
feat: 결제금액정보를 표시하는 기능 추가

semnil5202 committed:
refactor: 가격 정보에 대해 LocalScale 적용

semnil5202 committed:
docs: 누락된 기능구현 목록 추가 및 불필요한 기능 목록 제거

semnil5202 committed:
refactor: 상품 이미지가 없을 경우 대체 이미지를 보여주도록 변경

semnil5202 committed:
refactor: 불필요한 console.log 제거

semnil5202 committed:
refactor: 쿠폰 선택 Select font size 수정

semnil5202 committed:
refactor: 서버 URL 변경

semnil5202 committed:
fix: 사용한 쿠폰이 장바구니 페이지에서 선택 가능했던 오류 수정

igy95 commented on src/utils/checkNaturalStringTypeNumber.ts:
```diff
@@ -1,3 +1,3 @@
-export const isNaturalNumberString = (string: string) => {
+export const checkNaturalStringTypeNumber = (string: string) => {
```
정규식으로 좀 더 간단하게 처리해볼 수 있지 않을까요?

semnil5202 replied:
정규표현식으로 변경해보겠습니다!

igy95 commented on src/types/index.d.ts:
`.d.ts`에 선언하는 타입과 `.ts`에 선언하는 타입은 어떠한 기준으로 나누신 걸까요?

semnil5202 replied:
질문의 의도를 잘 이해하지 못했습니다. `.ts`에 선언한 타입이 어떤 것을 말씀하시는 걸까요?

`index.ts`는 따로 없어서 어떤 .ts를 말씀하시는지 잘 이해하지 못했습니다. 🙇

igy95 replied:
현재 앱 내에서 사용하기 위해 직접 선언해주신 타입과 인터페이스들은 모두 `d.ts` 파일에 들어가 있는데요, `.ts`파일에도 동일한 선언이 가능함에도 declaration file이 존재하는 이유와 사용 목적에 대해 찾아보시면 좋을 것 같습니다!

semnil5202 replied:
아하 단순히 `declare module '*.png';` 부분과 타입 및 인터페이스를 한 곳에서 사용하기 위해 위와 같이 작성했던 것 같습니다.

생각해보니 `d.ts`와 `.ts`의 차이점을 깊게 고려해보지 않은 것 같네요! 한 번 확인해보겠습니다. 감사합니다! 🙇

igy95 commented on src/layouts/SignUp.tsx:
```diff
@@ -0,0 +1,81 @@
+
+const isError = (error: any): error is Error => {
+  return error instanceof Error && typeof error.message === 'string';
+};
```
- `isError`는 특정 페이지 안에서만 사용되진 않을 것 같은데 선언 위치를 변경해야하지 않을까요?
- `error instanceof Error`에서 이미 `error is Error`가 보장되지 않나요?

semnil5202 replied:
네 isError가 SignUp 컴포넌트에서만 사용되는 것이 아니니 해당 함수는 분리될 필요성이 있어보입니다.

말씀해주신 것처럼 `typeof error.message === 'string'`은 불필요하였습니다. 실제로 사용해야하는 것이 error.message다 보니 착각한것 같습니다. 😅

현재는 promise.then.catch 문으로 변경하여 해당 함수는 불필요하여 제거하였습니다! 😀

igy95 commented on src/constants/index.ts:
```diff
@@ -0,0 +1,43 @@
+
+export const MOCK_URL = 'https://backend';
+
+export const SKELETONS_LENGTH = 8;
```
상수명이 꽤 포괄적이라는 느낌이 드네요! 이름만으로는 어떤 용도인지 구체적인 파악이 힘들어보입니다

semnil5202 replied:
PRODUCT 키워드를 앞에 붙여서 상품 관련 상수임을 좀 더 명확하게 하도록 하겠습니다!

igy95 commented on src/components/common/styles/Toast.styles.ts:
```diff
@@ -0,0 +1,48 @@
+import styled, { keyframes } from 'styled-components';
+import { ToastInfoType } from '../../../types';
+
+const toastAnim = keyframes`
```
`Anim`이라는 축약어는 조금 어색해보이네요 😅

semnil5202 replied:
grep에 검색을 해보았는데 Anim이란 명칭을 거의 안쓰네요 😅 풀네임으로 작성해보도록 하겠습니다

igy95 reviewed (changes requested):
안녕하세요 세인! 앱을 사용해보면서 회원가입, 로그인 UX가 추가되어 좀 더 현실의 앱을 사용하는 느낌이 들어 좋았네요. UX flow도 예상 범위 내에서 동작하는 것 같아 안정적으로 보였습니다.

### 전체 피드백

- 컴포넌트에서 전역 상태, api 조작에 대해 너무 상세히 알고 있다는 생각이 드는데요. 중간에 훅 레이어를 추가해서 컴포넌트 단에서 관심을 가져야할 범위를 줄여보는 게 어떨까요? 현재의 방향대로는 한 컴포넌트마다 비대한 로직을 담게 될 것으로 예상됩니다!

### QnA

> 상태가 점점 많아지고 복잡해지는 것 같은데, 리팩토링 방향을 어떻게 잡을지 고민입니다.

`atom` 폴더 내에서 `selector`도 같이 다루고 있는게 어색하게 느껴진다는 점을 제외하고는 현재의 관리 방법도 나쁘지 않아 보입니다. 오히려 전체 피드백에서 말씀 드린 것처럼 이것을 사용하는 쪽이 복잡해지는 것을 관리해야하지 않을까 생각이 드네요!

semnil5202 committed:
refactor: API 모듈 레이어 에러 헨들링 세분화

API 모듈 레이어에서 파악한 에러 종류는 다음과 같다.
1. fetch 하는 과정에서 발생하는 에러
2. response.ok 상태에 따른 에러

1번은 API 모듈 레이어 단에서 처리하기 복잡하며 코드 가독성이 전반적으로 떨어진다.
2번의 경우 서버에서 에러메세지가 주어지는 경우가 있고 없는 경우가 있다.
에러메세지 유무에 따라 에러를 정제하여 API 요청 레이어에서 세밀한 헨들링이 가능토록한다.

semnil5202 committed:
refactor: GET Products API 로직 useGetProductList 훅으로 분리

semnil5202 committed:
refactor: GET carts API 로직 useGetCartList 훅으로 분리

semnil5202 committed:
refactor: GET coupon API 로직 useGetCoupon 훅으로 분리

semnil5202 committed:
refactor: GET order API 로직 useGetOrderList 훅으로 분리

semnil5202 committed:
refactor: 불필요한 ``문 제거

semnil5202 committed:
fix: 서버에서 에러메시지가 있는 경우 Toast를 보여주지 못한 에러 수정

semnil5202 committed:
refactor: 오류가 났을 때 빈페이지 대신 Title은 보여주도록 변경

semnil5202 committed:
refactor: GET order API 로직 useGetOrderDetailList 훅으로 분리

semnil5202 committed:
fix(useGetOrderList): 잘못된 리턴 방식 수정

semnil5202 committed:
fix: 총 주문금액에서 배송비가 계산되지 않던 오류 수정

semnil5202 committed:
refactor: 불필요한 import 제거

semnil5202 committed:
refactor: Post cart API 로직 usePostCartItem 훅으로 분리

semnil5202 committed:
refactor: postCartItem async 오류 수정

semnil5202 committed:
refactor: cartItem API 매개변수 순서 수정 및 로직 수정

semnil5202 committed:
refactor: POST 장바구니 주문 API 로직 usePostPurchaseItem 훅으로 분리

semnil5202 committed:
refactor: 예기치 못한 server Error 발생 시 그에 맞는 오류 메세지 표시

semnil5202 committed:
refactor: POST signup API 로직 usePostSignUP 훅으로 분리

semnil5202 committed:
refactor: POST login API 로직 usePostLogin 훅으로 분리

semnil5202 committed:
refactor: DELETE cartItem API 로직 useDeleteCartItem 훅으로 분리

semnil5202 committed:
refactor: DELETE cartItems API 로직 useDeleteCartItems 훅으로 분리

semnil5202 committed:
refactor: 주문 목록 수량 및 총 가격 표시 방법 수정

semnil5202 committed:
refactor: font 색상 수정 (000 -> 333)

semnil5202 committed:
refactor: Image tag 공통 컴포넌트로 분리하여 중복 코드 제거

semnil5202 committed:
refactor: 로그인 상태에서는 서버를 바꿀 수 없도록 변경

서버마다 로그인 credential 상태 값이 모두 다르므로 이와 같이 변경한다.

semnil5202 committed:
refactor: 정규표현식으로 검증

semnil5202 committed:
feat: 아이디 및 비밀번호를 한글을 제외한 값으로 입력받을 수 있게 수정

semnil5202 committed:
refactor: 반응형 스타일 오류 수정

semnil5202 committed:
refactor: 불필요한 import 제거

semnil5202 committed:
refactor: 삼항연산자 대신 early return 적용

semnil5202 committed:
refactor: 포괄적인 상수명 보다 명확하게 변경

semnil5202 committed:
refactor: 변수명 수정 (Anim -> Animation)

semnil5202 commented:
안녕하세요 카일! 바쁘실텐데 빠르게 리뷰해주셔서 정말 감사드립니다! 😀👍

말씀해주셨던 API 로직 분리를 최우선적으로 생각하여 진행하였습니다. 추가로 API 통신 간 발생하는 에러를 모두 처리하였다고 생각했는데, 그렇지 못한 부분이 있어서 (CORS로 인한 Fetch 함수 자체에 대한 에러) 해당 부분 또한 보완하였습니다.

또한 서버에서 주는 에러 메세지를 최대한 활용하였지만, 간혹 서버에서 에러메세지를 전달하지 않는 경우도 있어, 그럴 경우 따로 에러메세지를 보여줄 수 있도록 처리하였습니다.

주요 리팩토링은 다음과 같습니다.

### 주요 리팩토링
- [x] API 로직 분리 (컴포넌트 -> 커스텀 훅)
- [x] 누락된 API 통신 에러 핸들링 보완
- [x] 일부 Page 반응형 오류 해결
- [x] Image 공통 컴포넌트로 분리 (onError 관련 로직 중복 제거 위함)
- [x] 로그인 상태에서 서버 변경 불가 처리 (서버마다 로그인 정보가 다르므로)
- [x] 로그인 및 회원가입 시 한글을 입력하지 못하도록 수정
- [x] 조건 렌더링이 크게 차이 날 경우 삼항연산자 대신 early return 적용

미리 리뷰 감사드립니다! 🙇

igy95 commented on src/components/hooks/usePostLogin.ts:
```diff
@@ -0,0 +1,37 @@
+  const setLoginState = useSetRecoilState(loginState);
+  const { showToast } = useToast();
+
+  const postLoginThroughApi = async (
+    serverName: ServerNameType,
+    payload: { name: string; password: string },
+    setIsModalOpen: React.Dispatch<SetStateAction<boolean>>
+  ) => {
```
다른 api hook도 마찬가지지만 해당 부분을 예시로 들자면, 훅을 제공하는 쪽과 사용하는 쪽 간에 너무 강하게 결합되어 있는게 아닌가 싶어요. api를 컴포넌트 단에 제공할 때 `postLoginThroughApi` 쪽에서 관심을 가져야할 부분은 다음과 같지 않을까요?

- 요청에 필요한 payload
- 요청 성공 / 실패에 따라 앱에서 후처리할 동작

`serverName`같은 경우도 사실 사용처가 고정되어 있다보니 인자로 주고받는 거보다 api 레이어에서 일찍 결정할 수 있었을 듯한데, 현재의 상황에서는 `usePostLogin` 내에서 `serverName`을 직접 가져다 쓰게 하면 최종적으로 사용하는 쪽에는 다음과 같은 인터페이스를 제공해볼 수 있을 것 같습니다.

```ts
// component.tsx

postLoginThroughApi(payload, {
  onSuccess: () => {},
  onError: () => {}
})
```

semnil5202 replied:
커스텀 훅으로 분리하면서 느꼈던 부분인데 좋은 피드백을 주셔서 감사드립니다!!

사실 serverName이란 상태를 두는 것이 맞는가도 고민을 해봤었는데, 로컬 스토리지에서 계속해서 get 해오는 것보단 상태로 두는게 낫다고 판단해서 상태로 둔 것이었습니다.

그리고 상태는 컴포넌트 단에서 가지고 있는 것이 맞다고 판단하여 지금의 구조가 되었네요.

말씀해주신 것처럼 관심사 분리가 잘 안 되어 추상화가 제대로 되지 못한 부분.. 앞으로도 계속 연습해보겠습니다. setIsModalOpen 함수 대신 onSuccess 로 받으면 결합도를 낮출 수 있겠군요 😀

좋은 피드백 정말 감사드립니다 카일! 🙇

igy95 reviewed (approved):
안녕하세요 세인~! 리뷰 반영하시느라 고생 많으셨습니다 :) 커밋 내역보니 리팩터링 하느라 고생 많았을 것 같네요 ㅎㅎ 추가적으로 2개 정도 코멘트 남겨두었으니 확인 부탁드리고 이번 미션은 이만 머지하도록 할게요 👍
